sample/watcher-timing.c

```diff
@@ -0,0 +1,315 @@
+	} else {
+		b = 2 * lhs_bin.count;
+		c = 2 * (lhs_total - needle);
+		z = (-b + sqrt(b * b - 4 * a * c)) / (2 * a);
+	}
+	return lhs_bin.centroid + (rhs_bin.centroid - lhs_bin.centroid) * z;
+}
+
+/**
+  This is an example of one way in which "prepare" and "check" watchers can be
+  useful. We track histograms of two timing metrics:
+
+  The first is "duration," which is the amount of time between a "check" and the
+  next "prepare" (in the next iteration of the event loop). This corresponds
+  pretty closely to the amount of time spent in event handlers (such as the
+  `on_timeout` handler in this example). In a real-world server, this would
+  provide a way to monitor whether any of your handlers are blocking or
+  otherwise performing heavy computation.
+
+  The second is "delay," which is the difference between the actual and expected
+  polling duration. The actual polling duration is the amount of time between a
+  "prepare" and the next "check" (in the same iteration of the event loop), and
+  the expected duration is obtained from `event_prepare_get_timeout`. In a
+  real-world server, this provides an indication of kernel scheduling delays.
+  For example, if your server is lightly loaded, this delay should usually be
+  close to your kernel's scheduling quantum (e.g. 1 millisecond).
+ */
+
+static struct timeval
+	prepare_time = { 0, 0 },
+	check_time = { 0, 0 },
+	expected = { 0, 0 };
+static struct histogram *durations, *delays;
+
+static void on_prepare(struct event_watcher *watcher, const struct event_prepare_cb_info *info)
+{
+	struct timeval duration;
+	evutil_gettimeofday(&prepare_time, NULL);
+	event_prepare_get_timeout(info, &expected);
+	if (check_time.tv_sec != 0) {
+		evutil_timersub(&prepare_time, &check_time, &duration);
+		histogram_update(durations, duration.tv_sec + duration.tv_usec / 1000000.0l);
+	}
+}
+
+static void on_check(struct event_watcher *watcher, const struct event_check_cb_info *info)
+{
+	struct timeval actual, delay;
+	evutil_gettimeofday(&check_time, NULL);
+	evutil_timersub(&check_time, &prepare_time, &actual);
+	evutil_timersub(&actual, &expected, &delay);
+	if (delay.tv_sec >= 0)
+		histogram_update(delays, delay.tv_sec + delay.tv_usec / 1000000.0l);
+}
+
+static void
+on_timeout(evutil_socket_t fd, short events, void *arg)
+{
+	printf("durations: [p50 = %fs, p95 = %fs], delays: [p50 = %fs, p95 = %fs]\n",
+		histogram_query(durations, 0.5),
+		histogram_query(durations, 0.95),
+		histogram_query(delays, 0.5),
+		histogram_query(delays, 0.95));
+}
+
+int
+main(int argc, char **argv)
+{
+	struct event_base *base = event_base_new();
+	struct event *ev = event_new(base, -1, EV_PERSIST, &on_timeout, NULL);
+	struct timeval one_second = { 1, 0 };
+
+	delays = histogram_new(100);
+	durations = histogram_new(100);
+
+	event_add(ev, &one_second);
+	event_watcher_prepare_new(base, &on_prepare);
+	event_watcher_check_new(base, &on_check);
+	event_base_dispatch(base);
+
```

Comment:
all below (`*free()`) will never be reached, but what you can do is add a handler for `SIGINT` (see `evsignal_*()`) and call `event_base_loopexit()` on it, then you will got no leaks on termination by interrupting.